File: sem1/hw8/p8-3/list.cpp
#include <iostream>
#include "list.h"

struct ListElement
{
    String *string;
    int value;
    ListElement *next;
};

struct List
{
    ListElement *head;
};


List *createList()
{
    List *newList = new List;
    newList->head = nullptr;
    return newList;
}

void addElementTo(List *list, String *string, int position)
{    
    if (position == 0)
    {
        ListElement *newElement = new ListElement;
        newElement->string = clone(string);
        newElement->next = list->head;
        newElement->value = 1;
        list->head = newElement;
        return;
    }

    ListElement *newElement = new ListElement;
    ListElement *temp = list->head;

    newElement->string = clone(string);
    newElement->value = 1;

    for (int i = 1; i < position; i++)
    {
        temp = temp->next;
    }

    newElement->next = temp->next;
    temp->next = newElement;
}

bool isEmpty(List *list)
{
    return (list->head == nullptr);
}

int sizeList(List *list)
{
    int result = 0;
    ListElement *temp = list->head;
    while (temp != nullptr)
    {
        result++;
        temp = temp->next;
    }
    return result;
}

void printList(List *list)
{

    ListElement *temp = list->head;

    int position = 0;
    while (temp != nullptr)
    {
        std::cout << position << ": string = ";
        printString(temp->string);
        std::cout << " " << "value = " << temp->value << "\n";
        position++;
        temp = temp->next;
    }
}

String *peekListElement(List *list, int position)
{
    ListElement *temp = list->head;
    for (int i = 0; i < position; i++)
    {
        temp = temp->next;
    }

    return temp->string;
}

void deleteElementFrom(List *list, int position)
{
    if (position == 0)
    {
        ListElement *temp = list->head;
        list->head = temp->next;
        deleteString(temp->string);
        delete temp;
        return;
    }

    ListElement *temp = list->head;
    for (int i = 0; i < position - 1; i++)
    {
        temp = temp->next;
    }

    ListElement *deleteThis = temp->next;
    temp->next = temp->next->next;

    deleteString(deleteThis->string);
    delete deleteThis;
}

bool isHere(List *list, String *string)
{
    ListElement *temp = list->head;

    while(temp != nullptr)
    {
        if (areEqual(string, temp->string))
            return true;
        temp = temp->next;
    }

    return false;
}

void increaseValue(List *list, String *string)
{
    ListElement *temp = list->head;

    while(temp != nullptr)
    {
        if (areEqual(string, temp->string))
        {
            temp->value++;
            return;
        }
        temp = temp->next;
    }
}

int getSumValues(List *list)
{
    int result = 0;

    if (isEmpty(list))
        return 0;

    ListElement *temp = list->head;
    while(temp != nullptr)
    {
        result = result + temp->value;
        temp = temp->next;
    }
    return result;
}

void deleteList(List *&list)
{
    while (!isEmpty(list))
        deleteElementFrom(list, 0);
    delete list;
}

File: sem2/p1-1/src/com/group144/kuzmin/List.java
package com.group144.kuzmin;

public class List {
    private ListElement head;
    private int size;

    public boolean isEmpty() {
        return size == 0;
    }

    public int size() {
        return size;
    }

    public void add(int value, int position) {
        if (position > size || position < 0)
            return;

        size++;
        if (position == 0) {
            head = new ListElement(value, head);
            return;
        }

        ListElement temp = head;

        for (int i = 1; i < position; i++)
            temp = temp.next;

        temp.next = new ListElement(value, temp.next);
    }

    public int peek(int position) {
        ListElement temp = head;

        for (int i = 0; i < position; i++)
            temp = temp.next;

        return temp.value;
    }

    public void delete(int position) {
        if (position >= size || position < 0)
            return;

        size--;
        if (position == 0) {
            head = head.next;
            return;
        }

        ListElement temp  = head;

        for (int i = 1; i < position; i++)
            temp = temp.next;

        temp.next = temp.next.next;

    }

    public void print() {
        ListElement temp = head;
        int i = 0;

        while (temp != null) {
            System.out.println("Pos: " + i + " value: " + temp.value);
            temp = temp.next;
            i++;
        }
    }

    private class ListElement {
        private int value;
        private ListElement next;

        ListElement(int value, ListElement next) {
            this.value = value;
            this.next = next;
        }
    }
}